sql/src/main/java/io/crate/cluster/gracefulstop/DecommissionAllocationDecider.java

```diff
@@ -0,0 +1,98 @@
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.cluster.gracefulstop;
+
+import com.google.common.collect.ImmutableSet;
+import io.crate.metadata.settings.CrateSettings;
+import org.elasticsearch.cluster.routing.RoutingNode;
+import org.elasticsearch.cluster.routing.ShardRouting;
+import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.cluster.routing.allocation.decider.AllocationDecider;
+import org.elasticsearch.cluster.routing.allocation.decider.Decision;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.node.settings.NodeSettingsService;
+
+import java.util.Map;
+import java.util.Set;
+
+
+@Singleton
+public class DecommissionAllocationDecider extends AllocationDecider {
+
+    public static final String NAME = "decommission";
+
+    private Set<String> decommissioningNodes = ImmutableSet.of();
+
+    private DataAvailability dataAvailability = DataAvailability.NONE;
+
+    @Inject
+    public DecommissionAllocationDecider(Settings settings, NodeSettingsService nodeSettingsService) {
+        super(settings);
+        ApplySettings applySettings = new ApplySettings();
+        applySettings.onRefreshSettings(settings);
+        nodeSettingsService.addListener(applySettings);
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (decommissioningNodes.contains(node.nodeId())
+            && dataAvailability == DataAvailability.PRIMARIES
+            && !shardRouting.primary()) {
+
+            // if primaries are removed from this node it will try to re-balance non-primaries onto this node
+            // prevent this - replicas that are already here can remain, but no new replicas should be assigned
+            return allocation.decision(Decision.NO, NAME, "dataAvailability=primaries, shard=replica, decommissioned=true");
+        }
+        return canRemainOrAllocate(node.nodeId(), shardRouting, allocation);
+    }
+
+    @Override
+    public Decision canRemain(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return canRemainOrAllocate(node.nodeId(), shardRouting, allocation);
+    }
+
+    private Decision canRemainOrAllocate(String nodeId, ShardRouting shardRouting, RoutingAllocation allocation) {
+        if (dataAvailability == DataAvailability.NONE) {
+            return allocation.decision(Decision.YES, NAME, "dataAvailability=none");
+        }
+
+        if (decommissioningNodes.contains(nodeId)) {
+            if (dataAvailability == DataAvailability.PRIMARIES && !shardRouting.primary()) {
+                return allocation.decision(Decision.YES, NAME, "dataAvailability=primaries shard=replica decommissioned=true");
```

Comment:
looks like the same logic like at canAllocate(), why not remove code from canAllocate() and always forward to this method?